utils/env_preferences/src/linux.rs

```diff
@@ -0,0 +1,112 @@
+                "LC_TELEPHONE" => Ok(LocaleCategory::Telephone),
+                "LC_MEASUREMENT" => Ok(LocaleCategory::Measurement),
+                "LC_IDENTIFICATION" => Ok(LocaleCategory::Identification),
+                "LC_ALL" => Ok(LocaleCategory::All),
+                _ => Ok(LocaleCategory::Other(s.to_string())),
+            }
+        }
+    }
+
+    pub fn get_locales() -> HashMap<LocaleCategory, String> {
+        let mut locale_map = HashMap::new();
+
+        // SAFETY: Safety is ensured because we pass a `NULL` pointer and retrieve the locale there is
+        // no subsequent calls for `setlocale` which could change the locale of this particular thread
+        let locales_ptr = unsafe { setlocale(LC_ALL, ptr::null()) };
+
+        if locales_ptr.is_null() {
+            locale_map.insert(LocaleCategory::All, "C".to_string());
+            return locale_map;
+        }
+
+        // SAFETY: A valid `NULL` terminator is present which is a requirement of `from_ptr`
+        let locales_cstr = unsafe { CStr::from_ptr(locales_ptr) };
+
+        if let Ok(locales_str) = locales_cstr.to_str() {
+            let locale_pairs = locales_str.split(';');
+
+            // To handle cases in case a single locale is returned or a list of locale
+            if locale_pairs.clone().count() == 1 {
+                locale_map.insert(LocaleCategory::All, "C".to_string());
+            } else {
+                for locale_pair in locale_pairs {
+                    let mut parts = locale_pair.split('=');
+                    if let (Some(key), Some(value)) = (parts.next(), parts.next()) {
+                        if let Ok(category) = LocaleCategory::from_str(key) {
+                            locale_map.insert(category, value.to_string());
+                        }
+                    }
+                }
+            }
+        }
+
+        locale_map
+    }
+
+    pub fn get_system_calendars() -> impl Iterator<Item = (Cow<'static, str>, Cow<'static, str>)> {
+        // SAFETY: Safety is ensured because we pass a `NULL` pointer and retrieve the locale there is
+        // no subsequent calls for `setlocale` which could change the locale of this particular thread
+        let locale_ptr = unsafe { setlocale(LC_TIME, ptr::null()) };
+
+        if !locale_ptr.is_null() {
+            // SAFETY: A valid `NULL` terminator is present which is a requirement of `from_ptr`
+            let c_str = unsafe { CStr::from_ptr(locale_ptr) };
+
+            if let Ok(str_slice) = c_str.to_str() {
+                // `gnome-calendar` is the default calendar and it only supports `Gregorian`.
+                // Related issue: https://gitlab.gnome.org/GNOME/gnome-calendar/-/issues/998
+                return Some((
+                    Cow::Owned(str_slice.to_string()),
+                    Cow::Borrowed("Gregorian"),
+                ))
+                .into_iter()
+                .chain(None);
+            }
```

Comment:
What is `gnome-calendar` here? How does it work on non-gnome Linux? Why are you adding `Gregorian`?

I'd just return locale without calendar here if your API on Linux doesn't  have any info on calendars. just return `fn get_system_calendars() -> Result<Vec<String>>` for locales and document that all locales are for gregorian calendar.

Reply:
I think the format should:

```rust
Result<String>
// LC_TIME -> only single locale can be returned
```